sdk/nodejs/runtime/resource.ts

```diff
@@ -273,6 +279,44 @@ export function readResource(res: Resource, parent: Resource | undefined, t: str
     }), label);
 }
 
+function getParentURN(parent?: Resource | Input<string>) {
+    if (Resource.isInstance(parent)) {
+        return parent.urn;
+    }
+    return output(parent);
+}
+
+function mapAliasesForRequest(aliases: (URN | Alias)[] | undefined, parentURN?: URN) {
+    if (aliases === undefined) {
+        return [];
+    }
+
+    return Promise.all(aliases.map(async a => {
+        const newAlias = new aliasproto.Alias();
+        if (typeof a === "string") {
+            newAlias.setUrn(a);
+        } else {
+            const newAliasSpec = new aliasproto.Alias.Spec();
+            const noParent = !a.hasOwnProperty("parent") && !parentURN;
+            newAliasSpec.setName(a.name);
+            newAliasSpec.setType(a.type);
+            newAliasSpec.setStack(a.stack);
+            newAliasSpec.setProject(a.project);
+            if (noParent) {
+                newAliasSpec.setNoparent(noParent);
+            } else {
+                const aliasParentUrn = a.hasOwnProperty("parent") ? getParentURN(a.parent) : output(parentURN);
+                const setUrn = aliasParentUrn.apply(urn => {
+                    newAliasSpec.setParenturn(urn);
+                });
+                all([setUrn]);
+            }
+            newAlias.setSpec(newAliasSpec);
+        }
+        return newAlias;
+    }));
+}
```

Comment:
@justinvp I've spent a good deal of time running through many permutations of implementations for this function.  This plus its call site in `registerResource` seem to be the crux of our failure in integration tests.  What we eventually want to do is call `newAliasSpec.setParenturn` on the urn of the parent passed into this alias.  I'd love any insight into how to access the value of an `Output` on a resource at this point in the code.

Reply:
@kpitzen Apologies for not replying sooner!

`Output<T>` has an internal `promise()` method that can be called to get the value (by awaiting it). Like this:

https://github.com/pulumi/pulumi/blob/45dddd89fe3dd82bae07fb46a6e79a9b99ccff91/sdk/nodejs/runtime/resource.ts#L526-L528